### DISCUSSION TITLE
[Outdated] Exmples from README of horde-ad (and more generally, the API for toying)

### DISCUSSION THREAD

**Discussion by Mikolaj** [Mikolaj/mostly-harmless#9] [General]
For now, this library does not implement the surface language from the paper nor the translation. It only implements the codomain of the translation, that is a set of combinators (an API), as compositionally, as feasible. That's a differentiation engine based on the formalism and algorithm in the paper. Of all such APIs, the first one implemented is a low-level one, a kernel library that may in the future be wrapped in something easier to use. Basic Haskell. Zero-cost abstractions used, exclusively.

At some point, it would be great to think how to make the API more user friendly and what performance trade-offs are acceptable for what gains in readability, generality, safety, etc.

In this discussions let's focus on short-term mprovements and understanding the perceived user needs, whether they are easy to fit into the current API, or not.

**Comment by awf** [comment]
Maybe a starting point is to work on a README for potential users of the library.  That user might expect to write code like
```haskell
-- [Please excuse Haskell errors, it's not a daily language for me]
f :: Autodiffable a => (a,a,a) -> (a,a)
f (x,y,z) = 
   let w = x * sin y
   in (atan2 (z,w), z * x)
```
and then, might hope to write something like
```haskell
let fval,(df1,df2) = $(diff f) (1.1,2.2,3.3)
in ...
```
I don't know how long a journey it is to such an interface, but it might work as a starting point?

**Reply by Mikolaj** [reply]
Oh, great, thank you for volunteering as the ~naive~ unspoiled user. Let me try to understand your expectations, while I try to meet them, step by step.

As a user, do you have Float or Double (or the C variants, CFloat or CDouble) or Complex (rather not, given the `atan2`?) in mind or are you really interested in running your program on more than one? Perhaps you are only making this effort at generality because you think the library requires it? In fact, the library fails when the example is applied to `[1.1, 2.2, 3.3]` due to not knowing what type of scalars to work with and the user is forced to declare the scalar type eventually. If it was declared right at the start, a lot of the constraints could have been omitted. 

The running example now looks as follows:

https://github.com/Mikolaj/mostly-harmless/blob/fde443b3dec7f08b593556945c2c9a85459931a8/test/Test.hs#L144-L197

Shall we try a monomorphic example? Or one with a repeated subexpression? Anything else?

Edit: Snippet updated: I've added a totally unsafe, but cute syntactic sugar for extracting the variables from the input vector.

Edit2: An updated version of this example is now in horde-ad README: https://github.com/Mikolaj/horde-ad/blob/master/README.md

**Reply by Mikolaj** [reply]
@awf: I've added a variant of this example that takes vector parameters and works on vectors whenever possible afterwards. Perhaps that's more what you had in mind? The initial part of the new code is quite wasteful, because the `Delta` operation `Index` (as envisioned in section 8.3) that I had to add allocates a new vector when evaluated. The original version, on scalars, did not allocate any vectors apart of the initial vector of parameters.

https://github.com/Mikolaj/horde-ad/blob/e7912a7b44b8174701bd59016e1d52b99cc26a3b/test/TestSingleGradient.hs#L178-L219

P.S. Sadly I don't know how to make github display code snippets from outside repos (horde-ad in this case).

**Comment by Mikolaj** [comment]
The examples in README have been updated (except for the QuickCheck tests of gradient vs derivative that should be generalized to arbitrary ds and dt (dt is currently set to ones) and extended to include the check that f (x + ds) is close to f x + f' x ds, the task is defined in https://github.com/Mikolaj/horde-ad/issues/27). 

Any comments please?

**Reply by Mikolaj** [reply]
@goldfirere, you mentioned wanting to link to the library during a talk in a near future. Would you have any comment about its README?
